File: engine/scalper.py
"""Scalper extras: Wilder-adjacent helpers, CVD divergence, 25bps book, EV bands."""
from __future__ import annotations

from typing import Dict, List, Optional, Tuple

from .util import to_float


def wilder_rsi(closes: List[float], period: int = 14) -> float:
    if len(closes) < period + 1:
        return 50.0
    deltas = [closes[i] - closes[i - 1] for i in range(1, len(closes))]
    gains = [max(d, 0.0) for d in deltas]
    losses = [max(-d, 0.0) for d in deltas]
    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period
    for g, loss in zip(gains[period:], losses[period:]):
        avg_gain = (avg_gain * (period - 1) + g) / period
        avg_loss = (avg_loss * (period - 1) + loss) / period
    if avg_gain == 0 and avg_loss == 0:
        return 50.0
    if avg_loss == 0:
        return 100.0
    if avg_gain == 0:
        return 0.0
    rs = avg_gain / avg_loss
    return 100.0 - (100.0 / (1.0 + rs))


def linear_slope(xs: List[float]) -> float:
    n = len(xs)
    if n < 2:
        return 0.0
    x_mean = sum(xs) / n
    y_mean = (n - 1) / 2
    num = sum((xs[i] - x_mean) * i for i in range(n))
    den = sum((i - y_mean) ** 2 for i in range(n))
    return num / den if den else 0.0


def detect_cvd_divergence(price_series: List[float], cvd_series: List[float], lookback: int = 20) -> dict:
    if len(price_series) < lookback or len(cvd_series) < lookback:
        return {"divergence": False, "type": None, "strength": 0.0}
    ps = linear_slope(price_series[-lookback:])
    cs = linear_slope(cvd_series[-lookback:])
    if ps > 0 and cs < 0:
        return {"divergence": True, "type": "bearish", "strength": abs(ps) + abs(cs)}
    if ps < 0 and cs > 0:
        return {"divergence": True, "type": "bullish", "strength": abs(ps) + abs(cs)}
    return {"divergence": False, "type": None, "strength": 0.0}


def build_cvd_series(taker_hist, current_price: float = 1.0) -> List[float]:
    if not isinstance(taker_hist, list) or not taker_hist:
        return []
    try:
        hist = sorted(taker_hist, key=lambda x: int(x.get("timestamp") or 0))
    except Exception:
        hist = taker_hist
    running = 0.0
    out = []
    for bar in hist:
        buy = to_float(bar.get("sumTakerBuyVolume") or bar.get("buyVol") or 0)
        sell = to_float(bar.get("sumTakerSellVolume") or bar.get("sellVol") or 0)
        running += (buy - sell) * current_price
        out.append(running)
    return out


def ev_and_band(momentum: dict, price: float) -> dict:
    biases, bands = {}, {}
    if price <= 0:
        return {"bias": biases, "bands": bands}
    for tf, mm in (momentum or {}).items():
        if not mm:
            continue
        rsi = mm.get("rsi", 50.0)
        ret_3 = mm.get("ret_3", 0.0)
        if rsi < 30:
            bias = ret_3 + 0.5
        elif rsi > 70:
            bias = ret_3 - 0.5
        else:
            bias = ret_3
        atr_pct = max(mm.get("atr_pct", 0.0), 0.0)
        half = (atr_pct / 100.0) * 1.5
        biases[tf] = bias
        bands[tf] = {
            "lower": price * (1.0 - half),
            "upper": price * (1.0 + half),
            "half_width_pct": atr_pct * 1.5,
        }
    return {"bias": biases, "bands": bands}


def depth_25bps(bids, asks, mid, bps: float = 25.0) -> Tuple[float, float, float]:
    if not mid or not bids or not asks:
        return 0.0, 0.0, 0.0
    lim = bps / 10000.0
    lo, hi = mid * (1.0 - lim), mid * (1.0 + lim)
    bid_d = sum(q for p, q in bids if p >= lo)
    ask_d = sum(q for p, q in asks if p <= hi)
    tot = bid_d + ask_d
    imb = (bid_d - ask_d) / tot * 100 if tot else 0.0
    return bid_d, ask_d, imb


def quality_weights(quality: Dict[str, float]) -> Dict[str, float]:
    w = {"OB": 20.0, "TF": 20.0, "OI": 15.0, "Funding": 10.0, "Liq": 15.0, "Vol": 10.0, "Mom": 10.0}
    mapping = {
        "order_book_quality": "OB", "trade_flow_quality": "TF", "oi_quality": "OI",
        "funding_quality": "Funding", "ls_ratio_quality": "Liq",
        "volume_quality": "Vol", "momentum_quality": "Mom",
    }
    for qk, fk in mapping.items():
        q = quality.get(qk, 50) / 100.0
        w[fk] = w[fk] * (0.5 + 0.5 * q)
    tot = sum(w.values()) or 1
    return {k: v / tot * 100 for k, v in w.items()}

File: android/app/src/main/java/com/coinglass/intel/domain/Scalper.kt
package com.coinglass.intel.domain

import com.coinglass.intel.domain.model.TakerBar
import kotlin.math.abs
import kotlin.math.max

object Scalper {
    fun wilderRsi(closes: List<Double>, period: Int = 14): Double {
        if (closes.size < period + 1) return 50.0
        val deltas = (1 until closes.size).map { closes[it] - closes[it - 1] }
        val gains = deltas.map { max(it, 0.0) }.toMutableList()
        val losses = deltas.map { max(-it, 0.0) }.toMutableList()
        var avgGain = gains.take(period).average()
        var avgLoss = losses.take(period).average()
        for (i in period until gains.size) {
            avgGain = (avgGain * (period - 1) + gains[i]) / period
            avgLoss = (avgLoss * (period - 1) + losses[i]) / period
        }
        if (avgGain == 0.0 && avgLoss == 0.0) return 50.0
        if (avgLoss == 0.0) return 100.0
        if (avgGain == 0.0) return 0.0
        val rs = avgGain / avgLoss
        return 100.0 - (100.0 / (1.0 + rs))
    }

    fun linearSlope(xs: List<Double>): Double {
        val n = xs.size
        if (n < 2) return 0.0
        val xMean = xs.average()
        val yMean = (n - 1) / 2.0
        var num = 0.0
        var den = 0.0
        for (i in 0 until n) {
            num += (xs[i] - xMean) * i
            den += (i - yMean) * (i - yMean)
        }
        return if (den == 0.0) 0.0 else num / den
    }

    fun volWeightedSlope(xs: List<Double>, vols: List<Double>): Double {
        val n = minOf(xs.size, vols.size)
        if (n < 3) return 0.0
        val x = xs.takeLast(n)
        val w = vols.takeLast(n).map { max(it, 1e-9) }
        val wSum = w.sum()
        val xMean = x.zip(w).sumOf { it.first * it.second } / wSum
        var iMean = 0.0
        for (i in 0 until n) iMean += i * w[i]
        iMean /= wSum
        var num = 0.0
        var den = 0.0
        for (i in 0 until n) {
            num += w[i] * (x[i] - xMean) * (i - iMean)
            den += w[i] * (i - iMean) * (i - iMean)
        }
        return if (den == 0.0) 0.0 else num / den
    }

    fun detectCvdDivergence(
        priceSeries: List<Double>,
        cvdSeries: List<Double>,
        volumes: List<Double> = emptyList(),
        lookback: Int = 20,
    ): Map<String, Any?> {
        if (priceSeries.size < lookback || cvdSeries.size < lookback) {
            return mapOf("divergence" to false, "type" to null, "strength" to 0.0)
        }
        val px = priceSeries.takeLast(lookback)
        val cv = cvdSeries.takeLast(lookback)
        val vs = if (volumes.size >= lookback) volumes.takeLast(lookback) else List(lookback) { 1.0 }
        val lastVol = vs.takeLast(5).average()
        val medVol = vs.sorted()[vs.size / 2]
        if (medVol > 0 && lastVol < medVol * 0.55) {
            return mapOf("divergence" to false, "type" to null, "strength" to 0.0, "filtered" to "low_volume")
        }
        val ps = volWeightedSlope(px, vs)
        val cs = volWeightedSlope(cv, vs)
        return when {
            ps > 0 && cs < 0 -> mapOf("divergence" to true, "type" to "bearish", "strength" to abs(ps) + abs(cs))
            ps < 0 && cs > 0 -> mapOf("divergence" to true, "type" to "bullish", "strength" to abs(ps) + abs(cs))
            else -> mapOf("divergence" to false, "type" to null, "strength" to 0.0)
        }
    }

    fun buildCvdSeries(takerHist: List<TakerBar>, currentPrice: Double = 1.0): List<Double> {
        if (takerHist.isEmpty()) return emptyList()
        val hist = takerHist.sortedBy { it.timestamp }
        var running = 0.0
        return hist.map { bar ->
            running += (bar.buyVol - bar.sellVol) * currentPrice
            running
        }
    }

    fun evAndBand(momentum: Map<String, Map<String, Double>>, price: Double): Pair<Map<String, Double>, Map<String, Pair<Double, Double>>> {
        val biases = mutableMapOf<String, Double>()
        val bands = mutableMapOf<String, Pair<Double, Double>>()
        if (price <= 0) return biases to bands
        for ((tf, mm) in momentum) {
            val rsi = mm["rsi"] ?: 50.0
            val ret3 = mm["ret_3"] ?: 0.0
            val bias = when {
                rsi < 30 -> ret3 + 0.5
                rsi > 70 -> ret3 - 0.5
                else -> ret3
            }
            val atrPct = max(mm["atr_pct"] ?: 0.0, 0.0)
            val half = (atrPct / 100.0) * 1.5
            biases[tf] = bias
            bands[tf] = (price * (1.0 - half)) to (price * (1.0 + half))
        }
        return biases to bands
    }

    fun depth25bps(
        bids: List<Pair<Double, Double>>,
        asks: List<Pair<Double, Double>>,
        mid: Double,
        bps: Double = 25.0,
    ): Triple<Double, Double, Double> {
        if (mid == 0.0 || bids.isEmpty() || asks.isEmpty()) return Triple(0.0, 0.0, 0.0)
        val lim = bps / 10_000.0
        val lo = mid * (1.0 - lim)
        val hi = mid * (1.0 + lim)
        val bidD = bids.filter { it.first >= lo }.sumOf { it.second }
        val askD = asks.filter { it.first <= hi }.sumOf { it.second }
        val tot = bidD + askD
        val imb = if (tot == 0.0) 0.0 else (bidD - askD) / tot * 100.0
        return Triple(bidD, askD, imb)
    }

    fun qualityWeights(quality: Map<String, Double>): MutableMap<String, Double> {
        val w = mutableMapOf(
            "OB" to 20.0, "TF" to 20.0, "OI" to 15.0, "Funding" to 10.0,
            "Liq" to 15.0, "Vol" to 10.0, "Mom" to 10.0,
        )
        val mapping = mapOf(
            "order_book_quality" to "OB",
            "trade_flow_quality" to "TF",
            "oi_quality" to "OI",
            "funding_quality" to "Funding",
            "ls_ratio_quality" to "Liq",
            "volume_quality" to "Vol",
            "momentum_quality" to "Mom",
        )
        for ((qk, fk) in mapping) {
            val q = (quality[qk] ?: 50.0) / 100.0
            w[fk] = w.getValue(fk) * (0.5 + 0.5 * q)
        }
        val tot = w.values.sum().let { if (it == 0.0) 1.0 else it }
        for (k in w.keys.toList()) w[k] = w.getValue(k) / tot * 100.0
        return w
    }
}
